File: python-services/tts_service.py
from flask import Flask, request, jsonify
import subprocess
import os
import time
import re
from pathlib import Path

app = Flask(__name__)

# Configuration
PIPER_PATH = '/home/techvoot/ai-tools/piper'
MODELS = {
    'hi': '/home/techvoot/ai-tools/piper/hi_IN-priyamvada-medium.onnx',
    'en': '/home/techvoot/ai-tools/piper/en_US-amy-medium.onnx'
}

def split_into_sentences(text, lang='en'):
    """Split text into sentences for chunked TTS"""
    if not text or not text.strip():
        return []
    
    # Clean up text
    text = text.strip()
    
    if lang == 'hi':
        # Hindi sentence endings: ।, ?, !
        # Split by sentence endings but keep the punctuation
        # Pattern: split on sentence endings (।, !, ?) optionally followed by space
        sentences = re.split(r'([।!?]+)\s*', text)
        result = []
        current = ''
        for part in sentences:
            if part.strip():
                current += part
                # Check if this part is a sentence ending
                if re.match(r'^[।!?]+$', part.strip()):
                    if current.strip():
                        result.append(current.strip())
                    current = ''
        if current.strip():
            result.append(current.strip())
        return [s for s in result if s.strip()]
    else:
        # English: split by . ! ? followed by space or end of string
        # Pattern: sentence ending punctuation followed by space or end
        sentences = re.split(r'([.!?]+)\s+', text)
        result = []
        current = ''
        
        for i, part in enumerate(sentences):
            if part.strip():
                current += part
                # If this part is only punctuation, it's a sentence ending
                if re.match(r'^[.!?]+$', part.strip()):
                    if current.strip():
                        result.append(current.strip())
                    current = ''
        
        # Add remaining text if any
        if current.strip():
            result.append(current.strip())
        
        # Fallback: if no sentences found (no punctuation), return whole text
        if not result:
            result = [text]
        
        return [s for s in result if s.strip()]

@app.route('/synthesize', methods=['POST'])
def synthesize():
    """Generate full audio (backward compatibility)"""
    try:
        data = request.json
        text = data.get('text')
        lang = data.get('lang', 'en')
        output_dir = data.get('output_dir')
        
        if not text:
            return jsonify({'success': False, 'error': 'Text is required'}), 400
        
        if lang not in MODELS:
            return jsonify({'success': False, 'error': f'Unsupported language: {lang}'}), 400
        
        timestamp = int(time.time() * 1000)
        output_file = os.path.join(output_dir, f'welcome_{timestamp}.wav')
        model_path = MODELS[lang]
        piper_binary = os.path.join(PIPER_PATH, 'build', 'piper')
        
        if not os.path.exists(piper_binary):
            return jsonify({'success': False, 'error': f'Piper binary not found: {piper_binary}'}), 500
        
        if not os.path.exists(model_path):
            return jsonify({'success': False, 'error': f'Model not found: {model_path}'}), 500
        
        process = subprocess.Popen(
            [
                piper_binary,
                '--model', model_path,
                '--length_scale', '0.80',
                '--noise_scale', '0.7',
                '--noise_w', '0.8',
                '--output_file', output_file
            ],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            cwd=PIPER_PATH,
            env={**os.environ, 'ESPEAK_DATA_PATH': '/usr/lib/x86_64-linux-gnu/espeak-ng-data'}
        )
        
        stdout, stderr = process.communicate(input=text)
        
        if process.returncode != 0:
            return jsonify({
                'success': False,
                'error': f'Piper failed: {stderr}'
            }), 500
        
        audio_url = f'/welcome_{timestamp}.wav'
        
        return jsonify({
            'success': True,
            'url': audio_url,
            'file_path': output_file
        })
        
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@app.route('/synthesize_chunk', methods=['POST'])
def synthesize_chunk():
    """Generate TTS for a single sentence/chunk"""
    try:
        data = request.json
        text = data.get('text')
        lang = data.get('lang', 'en')
        output_dir = data.get('output_dir')
        chunk_id = data.get('chunk_id', 0)
        
        if not text or not text.strip():
            return jsonify({'success': False, 'error': 'Text is required'}), 400
        
        if lang not in MODELS:
            return jsonify({'success': False, 'error': f'Unsupported language: {lang}'}), 400
        
        timestamp = int(time.time() * 1000)
        output_file = os.path.join(output_dir, f'chunk_{chunk_id}_{timestamp}.wav')
        model_path = MODELS[lang]
        piper_binary = os.path.join(PIPER_PATH, 'build', 'piper')
        
        if not os.path.exists(piper_binary):
            return jsonify({'success': False, 'error': f'Piper binary not found: {piper_binary}'}), 500
        
        if not os.path.exists(model_path):
            return jsonify({'success': False, 'error': f'Model not found: {model_path}'}), 500
        
        process = subprocess.Popen(
            [
                piper_binary,
                '--model', model_path,
                '--length_scale', '0.80',
                '--noise_scale', '0.7',
                '--noise_w', '0.8',
                '--output_file', output_file
            ],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            cwd=PIPER_PATH,
            env={**os.environ, 'ESPEAK_DATA_PATH': '/usr/lib/x86_64-linux-gnu/espeak-ng-data'}
        )
        
        stdout, stderr = process.communicate(input=text)
        
        if process.returncode != 0:
            return jsonify({
                'success': False,
                'error': f'Piper failed: {stderr}'
            }), 500
        
        audio_url = f'/chunk_{chunk_id}_{timestamp}.wav'
        
        return jsonify({
            'success': True,
            'url': audio_url,
            'file_path': output_file,
            'chunk_id': chunk_id,
            'text': text
        })
        
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@app.route('/split_sentences', methods=['POST'])
def split_sentences():
    """Split text into sentences for chunking"""
    try:
        data = request.json
        text = data.get('text')
        lang = data.get('lang', 'en')
        
        if not text:
            return jsonify({'success': False, 'error': 'Text is required'}), 400
        
        sentences = split_into_sentences(text, lang)
        
        return jsonify({
            'success': True,
            'sentences': sentences,
            'count': len(sentences)
        })
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@app.route('/health', methods=['GET'])
def health():
    return jsonify({'status': 'ok', 'service': 'tts'}), 200

if __name__ == '__main__':
    print('🎙️  Starting Python TTS Service on port 5002...')
    app.run(host='0.0.0.0', port=5002, debug=False)

File: services/ttsService.js
const axios = require('axios');
const path = require('path');

const TTS_SERVICE_URL = process.env.TTS_SERVICE_URL || 'http://localhost:5002';
const OUTPUT_DIR = path.join(__dirname, '..', 'public');

/**
 * Generate audio file using Python TTS service (full text - backward compatibility)
 * @param {string} text - Text to convert to speech
 * @param {string} lang - Language code (hi or en)
 * @returns {Promise<string>} - URL path to generated audio file
 */
async function generateWelcomeAudio(text, lang = 'en') {
  try {
    console.log(`🎙️  Calling Python TTS service...`);
    console.log(`   Language: ${lang}`);
    console.log(`   Text length: ${text.length} characters`);

    const response = await axios.post(`${TTS_SERVICE_URL}/synthesize`, {
      text: text,
      lang: lang,
      output_dir: OUTPUT_DIR
    }, {
      timeout: 30000 // 30 seconds timeout
    });

    if (response.data.success) {
      console.log(`✅ Audio generated: ${response.data.url}`);
      return response.data.url;
    } else {
      throw new Error(response.data.error || 'TTS service failed');
    }
  } catch (error) {
    console.error('❌ TTS Service Error:', error.message);
    // Fallback to old method if Python service is not available
    if (error.code === 'ECONNREFUSED' || error.code === 'ETIMEDOUT') {
      console.log('⚠️  Python TTS service not available, falling back to direct Piper call');
      return generateWelcomeAudioFallback(text, lang);
    }
    throw error;
  }
}

/**
 * Fallback: Generate audio using direct Piper call (old method)
 */
function generateWelcomeAudioFallback(text, lang = 'en') {
  const { spawn } = require('child_process');
  const fs = require('fs');
  const PIPER_CONFIG = require('../config/piper.config');
  const messages = require('../config/messages.json');

  return new Promise((resolve, reject) => {
    const timestamp = Date.now();
    const outputFile = path.join(__dirname, '..', 'public', `welcome_${timestamp}.wav`);
    const model = PIPER_CONFIG.modelPaths[lang];

    if (!model) {
      return reject(new Error(`${messages.errors.unsupportedLanguage}: ${lang}`));
    }

    console.log(`🎙️  ${messages.status.generatingAudio} (fallback mode)`);
    console.log(`   Language: ${lang}`);

    const piper = spawn(PIPER_CONFIG.piperBinary, [
      '--model', model,
      '--length_scale', PIPER_CONFIG.audioSettings.lengthScale.toString(),
      '--noise_scale', PIPER_CONFIG.audioSettings.noiseScale.toString(),
      '--noise_w', PIPER_CONFIG.audioSettings.noiseW.toString(),
      '--output_file', outputFile
    ], {
      cwd: PIPER_CONFIG.piperPath,
      env: {
        ...process.env,
        ESPEAK_DATA_PATH: PIPER_CONFIG.espeakDataPath
      }
    });

    piper.stdin.write(text);
    piper.stdin.end();

    let errorOutput = '';

    piper.stderr.on('data', (data) => {
      errorOutput += data.toString();
    });

    piper.on('close', (code) => {
      if (code !== 0) {
        console.error('❌ Piper error:', errorOutput);
        return reject(new Error(`Piper exited with code ${code}`));
      }

      const audioUrl = `/welcome_${timestamp}.wav`;
      console.log(`✅ ${messages.status.audioGenerated}: ${audioUrl}`);

      setTimeout(() => {
        fs.unlink(outputFile, (err) => {
          if (err) console.error('Error cleaning up audio file:', err);
          else console.log(`🗑️  ${messages.status.cleanedUp}: ${audioUrl}`);
        });
      }, PIPER_CONFIG.fileCleanupDelay);

      resolve(audioUrl);
    });

    piper.on('error', (error) => {
      console.error('❌ Failed to spawn Piper:', error);
      reject(error);
    });
  });
}

/**
 * Stream TTS: Split text into sentences and generate audio chunks
 * @param {string} text - Full response text
 * @param {string} lang - Language code
 * @param {Function} onChunkReady - Callback when each chunk is ready (chunkId, audioUrl, text)
 * @returns {Promise<void>}
 */
async function streamTTS(text, lang, onChunkReady) {
  try {
    console.log(`🎙️  Streaming TTS for text (${text.length} chars)...`);
    
    // Step 1: Split into sentences
    const splitResponse = await axios.post(`${TTS_SERVICE_URL}/split_sentences`, {
      text: text,
      lang: lang
    }, {
      timeout: 5000
    });

    if (!splitResponse.data.success) {
      throw new Error('Failed to split sentences');
    }

    const sentences = splitResponse.data.sentences;
    console.log(`→ Split into ${sentences.length} sentences`);

    if (sentences.length === 0) {
      throw new Error('No sentences found in text');
    }

    // Step 2: Generate TTS for each sentence
    const generateChunk = async (sentence, chunkId) => {
      const response = await axios.post(`${TTS_SERVICE_URL}/synthesize_chunk`, {
        text: sentence,
        lang: lang,
        output_dir: OUTPUT_DIR,
        chunk_id: chunkId
      }, {
        timeout: 20000
      });

      if (response.data.success) {
        return {
          chunkId: chunkId,
          url: response.data.url,
          text: sentence
        };
      } else {
        throw new Error(response.data.error);
      }
    };

    // Generate first chunk immediately (user hears it right away)
    console.log('→ Generating first chunk immediately...');
    const firstChunk = await generateChunk(sentences[0], 0);
    await onChunkReady(firstChunk.chunkId, firstChunk.url, firstChunk.text);
    console.log(`✅ First chunk ready: ${firstChunk.url}`);

    // Generate remaining chunks in parallel (faster)
    if (sentences.length > 1) {
      console.log(`→ Generating remaining ${sentences.length - 1} chunks in parallel...`);
      const remainingPromises = sentences.slice(1).map((sentence, index) => 
        generateChunk(sentence, index + 1)
      );

      // Process chunks as they complete (stream them)
      for (const promise of remainingPromises) {
        const chunk = await promise;
        await onChunkReady(chunk.chunkId, chunk.url, chunk.text);
        console.log(`✅ Chunk ${chunk.chunkId} ready: ${chunk.url}`);
      }
    }

    console.log('✅ All TTS chunks generated');
  } catch (error) {
    console.error('❌ Stream TTS Error:', error.message);
    if (error.code === 'ECONNREFUSED' || error.code === 'ETIMEDOUT') {
      console.log('⚠️  Python TTS service not available, using fallback');
      // Fallback: generate full audio
      const fullAudioUrl = await generateWelcomeAudioFallback(text, lang);
      await onChunkReady(0, fullAudioUrl, text);
    } else {
      throw error;
    }
  }
}

module.exports = { 
  generateWelcomeAudio,
  streamTTS 
};
